nemo/collections/asr/models/k2_sequence_models.py

```diff
@@ -0,0 +1,393 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional
+
+import k2
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+
+from nemo.collections.asr.losses.lattice_losses import LatticeLoss
+from nemo.collections.asr.models.ctc_bpe_models import EncDecCTCModelBPE
+from nemo.collections.asr.models.ctc_models import EncDecCTCModel
+from nemo.collections.asr.modules.graph_decoder import ViterbiDecoderWithGraph
+from nemo.collections.asr.parts.k2.utils import load_graph
+from nemo.core.classes.common import PretrainedModelInfo, typecheck
+from nemo.utils import logging
+
+
+class EncDecK2SeqModel(EncDecCTCModel):
+    """Encoder decoder models with various lattice losses."""
+
+    @classmethod
+    def list_available_models(cls) -> Optional[PretrainedModelInfo]:
+        """
+        This method returns a list of pre-trained model which can be instantiated directly from NVIDIA's NGC cloud.
+
+        Returns:
+            List of available pre-trained models.
+        """
+        pass
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        loss_kwargs = self._cfg.get("loss", {})
+
+        # collecting prior knowledge for MAPLoss
+        self.use_graph_lm = loss_kwargs.get("criterion_type", "ml") == "map"
+        if self.use_graph_lm:
+            self.token_lm = None
+            self.token_lm_cache_dict = None
+            self.token_lm_path = self._cfg.get("token_lm", None)
+            token_lm_overwrite = self._cfg.get("token_lm_overwrite", False)
+            if token_lm_overwrite:
+                logging.info(
+                    f"""Overwriting token_lm with `{self.token_lm_path}`. 
+                             Previously saved token_lm, if it exists, will be ignored."""
+                )
+                self.token_lm = load_graph(self.token_lm_path)
+                loss_kwargs["token_lm"] = self.token_lm
+
+        self._update_k2_modules(loss_kwargs)
+
+    def _update_k2_modules(self, loss_kwargs):
+        """
+        Helper function to initialize or update k2 loss and transcribe_decoder.
+        """
+        del self.loss
+        if hasattr(self, "transcribe_decoder"):
+            del self.transcribe_decoder
+
+        self.loss = LatticeLoss(
+            num_classes=self.decoder.num_classes_with_blank - 1,
+            reduction=self._cfg.get("ctc_reduction", "mean_batch"),
+            **loss_kwargs,
+        )
+        remove_consecutive = loss_kwargs.get("topo_with_selfloops", True) and loss_kwargs.get(
+            "topo_type", "default"
+        ) not in ["forced_blank", "identity",]
+        self._wer.remove_consecutive = remove_consecutive
+
+        criterion_type = self.loss.criterion_type
+        transcribe_training = self._cfg.get("transcribe_training", False)
```

Comment:
This arg is for the model to recompute argmax values with graph decoding in case of the MMI loss. We don't specifically need that in training, so this parameter is about that.